pkg/eestream/encode.go

```diff
@@ -38,71 +38,103 @@ type ErasureScheme interface {
 }
 
 type encodedReader struct {
-	r               io.Reader
-	es              ErasureScheme
-	cv              *sync.Cond
-	inbuf           []byte
-	outbufs         [][]byte
-	piecesRemaining int
-	err             error
+	r       io.Reader
+	es      ErasureScheme
+	inbuf   []byte
+	outbufs [][]byte
+	chans   map[int]chan []byte
+	err     error
 }
 
 // EncodeReader will take a Reader and an ErasureScheme and return a slice of
 // Readers
-func EncodeReader(r io.Reader, es ErasureScheme) []io.Reader {
+func EncodeReader(r io.Reader, es ErasureScheme, maxBufferMemory int) []io.Reader {
 	er := &encodedReader{
 		r:       r,
 		es:      es,
-		cv:      sync.NewCond(&sync.Mutex{}),
 		inbuf:   make([]byte, es.DecodedBlockSize()),
 		outbufs: make([][]byte, es.TotalCount()),
+		chans:   make(map[int]chan []byte, es.TotalCount()),
 	}
 	readers := make([]io.Reader, 0, es.TotalCount())
 	for i := 0; i < es.TotalCount(); i++ {
-		er.outbufs[i] = make([]byte, 0, es.EncodedBlockSize())
 		readers = append(readers, &encodedPiece{
 			er: er,
 			i:  i,
 		})
 	}
+	if maxBufferMemory < 0 {
+		er.err = Error.New("negative max buffer memory")
+		return readers
+	}
+	chanSize := maxBufferMemory / (es.TotalCount() * es.EncodedBlockSize())
+	if chanSize < 1 {
+		chanSize = 1
+	}
+	for i := 0; i < es.TotalCount(); i++ {
+		er.chans[i] = make(chan []byte, chanSize)
+	}
+	go er.fillBuffer()
 	return readers
 }
 
-func (er *encodedReader) wait() (err error) {
-	// have we already failed? just return that
-	if er.err != nil {
-		return er.err
+func (er *encodedReader) fillBuffer() {
+	defer er.closeChannels()
+	for {
+		_, err := io.ReadFull(er.r, er.inbuf)
+		if err != nil {
+			return
+		}
+		err = er.es.Encode(er.inbuf, er.addToReader)
+		if err != nil {
+			return
+		}
+	}
+}
+
+func (er *encodedReader) addToReader(num int, data []byte) {
+	if er.chans[num] == nil {
+		// this channel is already closed for slowliness - skip it
+		return
 	}
-	// are other pieces still using buffer? wait on a condition variable for
-	// the last remaining piece to fill all the buffers.
-	if er.piecesRemaining > 0 {
-		er.cv.Wait()
-		// whoever broadcast a wakeup either set an error or filled the buffers.
-		// er.err might be nil, which means the buffers are filled.
-		return er.err
+	// data is reused by infecious, so add a copy to the channel
+	tmp := make([]byte, len(data))
+	copy(tmp, data)
+	// add data copy to the respective reader's channel
+	for {
+		select {
+		case er.chans[num] <- tmp:
+			return
+		// block for no more than 50 ms
+		case <-time.After(50 * time.Millisecond):
```

Comment:
since these 50ms sleeps are all happening serially during the rs encode, does it make sense to kick them all off in parallel? essentially, put the for loop in a goroutine? we copied data already anyway so that should be safe to return without the for loop having finished

Reply:
also, what happens if a channel is reliably taking 49ms every write? i kind of wonder if it makes sense to just try a non-blocking send on er.chans[num] and check if the channel is slow immediately.

one last thing - time.After is kind of scary. no matter what happens we're kicking off a goroutine that sleeps for 50ms, every for loop iteration, which is not garbage collected if we exit early

Reply:
after thinking about this more, I think you should make addToReader an inline function inside of fillBuffer. It will be easier to kick off a bunch of parallel writes to er.chans and then wait for them all to either finish or time out.

Reply:
Thanks for the comment about `time.After()`. I didn't know that and it should be fixed for sure.

I don't think that using just non-blocking send would work well. It would be the common case that the input is faster than the output. If the input is faster it will hit full channels very often (if not all the time) and it will burn the CPU. I really hate relying on such 50 ms timeouts, but so far it's the best I figured out to avoid the CPU burn in this case.

I'll try the idea for parallel channel send.